src/webgpu/api/operation/vertex_state.spec.ts

```diff
@@ -0,0 +1,348 @@
+          (col === row || col + row === kWidth - 1)
+        ) {
+          texelValueBytes = new Uint8Array([0x00, 0xff, 0x00, 0xff]);
+        } else {
+          texelValueBytes = new Uint8Array([0x00, 0x00, 0x00, 0xff]);
+        }
+
+        const byteOffset = row * this.bytesPerRow + col * texelValueBytes.byteLength;
+        arrayBuffer.set(texelValueBytes, byteOffset);
+      }
+    }
+    return arrayBuffer;
+  }
+}
+
+export const g = makeTestGroup(IndexFormatTest);
+
+// Test indexing draw with index format of uint16. If this is interpreted as uint32, it will have
+// index 1 and 2 be both 0 and render nothing. And the index buffer size - offset must be not less
+// than the size required by triangle list, otherwise it also render nothing.
+g.test('index_format_uint16')
+  .params(params().combine(poptions('indexOffset', [0, 6, 18])))
+  .fn(t => {
+    const { indexOffset } = t.params;
+
+    const indexArray = new Uint16Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint16');
+    const indexCount = indexArray.length;
+    t.run(indexBuffer, indexCount, 'uint16', indexOffset);
+
+    let expectedShape: RenderShape;
+    if (indexOffset === 0) {
+      expectedShape = RenderShape.Square;
+    } else if (indexOffset === 6) {
+      expectedShape = RenderShape.BottomLeftTriangle;
+    } else {
+      expectedShape = RenderShape.Nothing;
+    }
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(expectedShape);
+    t.expectContents(t.result, expectedTextureValues);
+  });
+
+// Test indexing draw with index format of uint32. If this is interpreted as uint16, then it would
+// be 0, 1, 0, ... and would draw nothing. And the index buffer size - offset must be not less than
+// the size required by triangle list, otherwise it also render nothing.
+g.test('index_format_uint32')
+  .params(params().combine(poptions('indexOffset', [0, 12, 36])))
+  .fn(t => {
+    const { indexOffset } = t.params;
+
+    const indexArray = new Uint32Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint32');
+    const indexCount = indexArray.length;
+    t.run(indexBuffer, indexCount, 'uint32', indexOffset);
+
+    let expectedShape: RenderShape;
+    if (indexOffset === 0) {
+      expectedShape = RenderShape.Square;
+    } else if (indexOffset === 12) {
+      expectedShape = RenderShape.BottomLeftTriangle;
+    } else {
+      expectedShape = RenderShape.Nothing;
+    }
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(expectedShape);
+    t.expectContents(t.result, expectedTextureValues);
+  });
+
+// Test primitive restart with each primitive topology. The primitive restart value can used with
+// strip primitive topologies ('line-strip' or 'triangle-strip').
+// For line-strip, [0, 1, PRIM_RESTART, 2, 3] only render a X shape.
+// For triangle-strip, [0, 1, 3, PRIM_RESTART, 2] only the first triangle.
```

Comment:
This description of these index arrays looks incomplete relative to the code. Can just delete this part of the comment.